### ISSUE TITLE
Storing Same Owned Entity In Json Column Does Not Work

### ISSUE THREAD

**Issue by dradovic** [dotnet/efcore#33967]
## Description

When the same instance of an object is referenced from multiple entities, it is only stored once and silently discarded in other parents when storing it in a JSON column. IMHO, at least an error should be thrown instead of just dropping the data if for some reason it is not possible to store a JSON serialized copy of "shared" instances or if that would lead to an undesirable design.

### How to reproduce - the code

In order for you to reproduce easily, I've extracted the problem to this repository:
https://github.com/dradovic/MissingLocationJsonRepro

The repo contains 3 succeedingly stripped down cases in 3 branches.

1. `main` branch: The story starts with an entity called `Boat`, where `Passengers` are stored in a JSON column.
```
public class Boat
{
    public int Id { get; set; }
    public ICollection<PassengerInfo> Passengers { get; set; } = null!;
    public override string ToString() => $"Boat {Id}: {string.Join(" ; ", Passengers)}";
}
...
builder.Entity<Boat>().OwnsMany(b => b.Passengers).ToJson();
```
A `PassengerInfo` is reflecting some Person data:
```
public class PassengerInfo
{
    public string Name { get; set; } = null!;
    public Location Location { get; set; } = null!;
    public override string ToString() => $"Passenger: {Name} @ {Location}";
}

public class Location
{
    public string Address1 { get; set; } = null!;
    public string City { get; set; } = null!;
    public override string ToString() => $"{Address1} in {City}";
}
```
Populating some boats with passengers whose `Location` points to the *same* instance results in:
```
foreach (var boat in db.Boats)
{
    Console.WriteLine(boat);
    // Boat 5: Passenger: Joe Doe @
    // Boat 6: Passenger: Joe Doe @
    // Boat 7: Passenger: Joe Doe @ Jumpy Rd. 1 in Duville
    // Id	Passengers
    // 5    [{ "Name":"Joe Doe","Location":null}]
    // 6    [{ "Name":"Joe Doe","Location":null}]
    // 7    [{ "Name":"Joe Doe","Location":{ "Address1":"Jumpy Rd. 1","City":"Duville"} }]
}
```
As we can see, the location is only stored for the last boat's passenger but silently dropped for the others(!).

2. `OnlyLocation` branch: I then went on to remove the intermediary `PassangerInfo` class and to try to store the locations directly.
```
public class Boat
{
    public int Id { get; set; }
    public ICollection<Location> Passengers { get; set; } = null!;
    public override string ToString() => $"Boat {Id}: {string.Join(" ; ", Passengers)}";
}
``` 
Again, data is silently discarded if we try to store the same location in multiple rows:
```
foreach (var boat in db.Boats)
{
    Console.WriteLine(boat);
    // Boat 8:
    // Boat 9:
    // Boat 10: Jumpy Rd. 1 in Duville
    // Id	Passengers
    // 8    []
    // 9    []
    // 10   [{ "Address1":"Jumpy Rd. 1","City":"Duville"}]
}
```
This time, JSON serialization does not say `null` but just discards the whole object.

3. `OnlyOneLocation` branch: in this branch, I've changed the collection to be a one-to-one relationship to `Location`.
```
public class Boat
{
    public int Id { get; set; }
    public Location Location { get; set; } = null!;
    public override string ToString() => $"Boat {Id}: {Location}";
}
```
And this time, the data cannot even be persisted as I get a: `SqlException: Cannot insert the value NULL into column 'Passenger', table 'MissingLocationJsonRepro.dbo.Boats'` if I try to persist the same location referenced from multiple boats which shows that EF again tried to drop data.

### Include provider and version information

EF Core version:
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: Windows
IDE: Visual Studio 2022 17.10.1

**Comment by ajcvickers** [comment]
> Populating some boats with passengers whose Location points to the same instance

When the code move the same instance like this, it is changing the relationship of that single instance. So it is first associated with passenger 1, then removed from passenger 1 and associated with passenger 2, and so on. A single object instance cannot be used to represent multiple different entities. This is one of the reasons why [complex types](https://github.com/dotnet/efcore/issues/31252) are a better fit than entity types for mapping to JSON.

**Comment by dradovic** [comment]
@ajcvickers I see. I think that the confusing part is that `Location` is used as an *owned* entity throughout our app. (The model is configured with `builder.Owned<Location>();`). And so, `Location` has no `Id` or key or whatsoever that would imply a sense of referential *identity* on a DB-level. Which in turn implies a by-copy/value semantic. Even more so, when that object just needs to be dumped into a JSON string.
So, if I have an object graph where here and there a `location` is being referenced, I would have expected it to be just inlined into the JSON without any "identity tracking". Certainly not that it only appears in one place and silently is discarded in all other places.

Reading your [EF Core 8 RC1: Complex types as value objects](https://devblogs.microsoft.com/dotnet/announcing-ef8-rc1/) makes it clear to me that we should switch the modelling of `Location` to rather be a Complex Type than an Owned Type. However, this will then make us dependent on #31252 (if we want to store locations in JSON) if my understanding is correct.

There's also #32799 which to a large extent seems to be a duplicate of this issue here.

So, if you don't plan on any changes - even throwing an error, which seems to be hard to implement according to your comment [here](https://github.com/dotnet/efcore/issues/32799#issuecomment-1903722216) - we can close this issue.

Thanks a lot for your valuable input.

**Comment by ajcvickers** [comment]
> Which in turn implies a by-copy/value semantic. Even more so, when that object just needs to be dumped into a JSON string.

Unfortunately, this is not how owned types work. They are still entity types and follow entity semantics. 

> we should switch the modelling of Location to rather be a Complex Type than an Owned Type.

Correct.

> However, this will then make us dependent on https://github.com/dotnet/efcore/issues/31252 (if we want to store locations in JSON) if my understanding is correct.

Yes, this is not a great place for us to be in, but prioritization of resources has been very difficult. We're working hard to get to a consistent space.
